scylla/src/transport/cluster.rs

```diff
@@ -0,0 +1,227 @@
+
+        let (refresh_sender, refresh_receiver) = tokio::sync::mpsc::channel(32);
+
+        let worker = ClusterWorker {
+            cluster_data: cluster_data.clone(),
+
+            topology_reader: TopologyReader::new(initial_peers, compression),
+            compression,
+
+            refresh_channel: refresh_receiver,
+        };
+
+        let (fut, worker_handle) = worker.work().remote_handle();
+        tokio::spawn(fut);
+
+        let result = Cluster {
+            data: cluster_data,
+            refresh_channel: refresh_sender,
+            _worker_handle: worker_handle,
+        };
+
+        result.refresh_topology().await?;
+
+        Ok(result)
+    }
+
+    pub fn get_data(&self) -> Arc<ClusterData> {
+        self.data.read().unwrap().clone()
+    }
+
+    pub async fn refresh_topology(&self) -> Result<(), QueryError> {
+        let (response_sender, response_receiver) = tokio::sync::oneshot::channel();
+
+        self.refresh_channel
+            .send(RefreshRequest {
+                response_chan: response_sender,
+            })
+            .await
+            .unwrap_or_else(|_| panic!("Bug in Cluster::refresh_topology sending"));
+        // Other end of this channel is in ClusterWorker, can't be dropped while we have &self to Cluster with _worker_handle
+
+        response_receiver
+            .await
+            .unwrap_or_else(|_| panic!("Bug in Cluster::refresh_topology receiving"))
+        // ClusterWorker always responds
+    }
+
+    /// Returns nonempty list of working connections to all shards
+    pub async fn get_working_connections(&self) -> Result<Vec<Arc<Connection>>, QueryError> {
+        let cluster_data: Arc<ClusterData> = self.get_data();
+        let peers = &cluster_data.known_peers;
+
+        let mut result: Vec<Arc<Connection>> = Vec::with_capacity(peers.len());
+
+        let mut last_error: Option<QueryError> = None;
+
+        for node in peers.values() {
+            match &*node.connections.read().await {
+                NodeConnections::Single(conn_keeper) => match conn_keeper.get_connection().await {
+                    Ok(conn) => result.push(conn),
+                    Err(e) => last_error = Some(e),
+                },
+                NodeConnections::Sharded { shard_conns, .. } => {
+                    for conn_keeper in shard_conns.values() {
+                        match conn_keeper.get_connection().await {
+                            Ok(conn) => result.push(conn),
+                            Err(e) => last_error = Some(e),
+                        }
+                    }
+                }
+            }
+        }
+
+        if result.is_empty() {
+            return Err(last_error.unwrap()); // By invariant peers is nonempty
+        }
+
+        Ok(result)
+    }
+}
+
+impl ClusterWorker {
+    pub async fn work(mut self) {
+        use tokio::time::{Duration, Instant};
+
+        let refresh_duration = Duration::from_secs(60); // Refresh topology every 60 seconds
```

Comment:
I suppose it would be nice to make this parameter configurable eventually. And to allow triggering the topology sync on demand when users want it, e.g. because they are aware that topology just changed and they don't want to wait 60 seconds for the changes to be applied. But it's good as is for now.

Reply:
There is already an option to refresh on demand - there is `Session::refresh_topology` which calls `Cluster::refresh_topology`.

About configuration I think it can be easily added later after we add multiple thinks to configure (TLS, password, load balancing policy etc..) which would require some `SessionBuilder` or something like that.

Reply:
Oh, right, it used to be manual only at first.